File: 20131209/AM/131209_tests/swallow/solutions/memoised.java
import java.io.*;
import java.util.*;

public class swallow {
  public static final int MAX_SIZE = 1000;

  public static String folded[][] = new String[MAX_SIZE+1][MAX_SIZE+1];

  public static void fold(String s, int i, int j) {
    if (folded[i][j] != null)
      return;

    folded[i][j] = s.substring(i, j);
    if (j-i == 1) {
      return;
    }

    for (int k = i+1; k < j; ++k) {
      fold(s, i, k);
      fold(s, k, j);

      String p1a = folded[i][k];
      String p2a = folded[k][j];

      if (p1a.length() + p2a.length() < folded[i][j].length())
        folded[i][j] = p1a + p2a;

      if ((j-k) % (k-i) == 0) {
        final String p1 = s.substring(i, k);
        int idx = k;
        boolean is_repeated = true;

        while (idx != j) {
          if (!s.substring(idx, idx+p1.length()).equals(p1)) {
            is_repeated = false;
            break;
          }
          idx += p1.length();
        }

        if (is_repeated) {
          String t = "" + ((j-k)/p1.length()+1);
          if (t.length() + p1a.length() + 2 < folded[i][j].length()) {
            folded[i][j] = t + "(" + p1a + ")";
          }
        }
      }
    }
  }

  public static void main(String[] args) throws Exception {
    BufferedReader fin = new BufferedReader(new InputStreamReader(System.in));
    int N = Integer.parseInt(fin.readLine());
    String s = fin.readLine().trim();

    fold(s, 0, s.length());
    
    System.out.println(folded[0][s.length()]);
  }
}

File: 20131209/AM/131209_tests/swallow/solutions/memoised.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <map>
#include <vector>
#include <utility>

using namespace std;

string toString(int v) {
  ostringstream oss;
  oss << v;

  return oss.str();
}

map<string, string> folded;

void fold(const string& s) {
  if (folded.count(s) != 0)
    return;

  folded[s] = s;
  //repeats[s][s] = 1;

  for (int i = 1; i < s.size(); ++i) {
    const string p1 = s.substr(0, i);
    const string p2 = s.substr(i);
    fold(p1);
    fold(p2);
    const string p1a = folded[p1];
    const string p2a = folded[p2];

    string t = p1a + p2a;

    if (t.size() < folded[s].size())
      folded[s] = t;

    if (p2.size() % p1.size() == 0) {
      int idx = 0;
      bool is_repeated = true;

      while (idx != p2.size()) {
        if (p2.substr(idx, p1.size()) != p1) {
          is_repeated = false;
          break;
        }
        idx += p1.size();
      }

      if (is_repeated) {
        t = toString(p2.size()/p1.size()+1) + "(" + p1a + ")";
        if (t.size() < folded[s].size()) {
          folded[s] = t;
        }
      }
    }
  }
}

int main(int argc, char** argv) {
  int N;
  cin >> N;
  string s;
  cin >> s;

  fold(s);
  cout << folded[s] << endl;

  return 0;
}
